src/main/kotlin/com/exactpro/th2/check1/rule/AbstractCheckTask.kt

```diff
@@ -469,22 +552,61 @@ abstract class AbstractCheckTask(
-            .filter { message -> message.metadata.id.sequence > sequence }
+            .filter { message ->
+                if (message.metadata.id.sequence == sequence) {
+                    bufferContainsStartMessage = true
+                }
+                message.metadata.id.sequence > sequence
+            }
 
-    private fun Checkpoint.getSequence(sessionKey: SessionKey): Long {
-        val sequence = sessionAliasToDirectionCheckpointMap[sessionKey.sessionAlias]
-            ?.directionToSequenceMap?.get(sessionKey.direction.number)
+    private fun Checkpoint.getCheckpointData(sessionKey: SessionKey): CheckpointData {
+        val checkpointData = sessionAliasToDirectionCheckpointMap[sessionKey.sessionAlias]
+            ?.directionToCheckpointDataMap?.get(sessionKey.direction.number)
 
-        if (sequence == null) {
+        if (checkpointData == null) {
             if (LOGGER.isWarnEnabled) {
-                LOGGER.warn("Checkpoint '{}' doesn't contain sequence for session '{}'", shortDebugString(this), sessionKey)
+                LOGGER.warn("Checkpoint '{}' doesn't contain checkpoint data for session '{}'", this.toJson(), sessionKey)
             }
-        } else {
+            val sequence = sessionAliasToDirectionCheckpointMap[sessionKey.sessionAlias]
+                ?.directionToSequenceMap?.get(sessionKey.direction.number)
+            if (sequence == null) {
+                if (LOGGER.isWarnEnabled) {
+                    LOGGER.warn("Checkpoint '{}' doesn't contain sequence for session '{}'", this.toJson(), sessionKey)
+                }
+                return CheckpointData(DEFAULT_SEQUENCE)
+            }
+            return CheckpointData(sequence)
+        }
+
+        return checkpointData.convert().apply {
             LOGGER.info("Use sequence '{}' from checkpoint for session '{}'", sequence, sessionKey)
         }
+    }
 
-        return sequence ?: DEFAULT_SEQUENCE
+    private fun checkOnMessageTimeout(timestamp: Timestamp): Boolean {
+        return checkpointTimeout == null || checkpointTimeout!!.isAfter(timestamp) || checkpointTimeout == timestamp
     }
 
-    private data class Legacy(val executorService: ExecutorService, val lastSequence: Long)
+    /**
+     * Provides the ability to stop observing if a message timeout is set.
+     */
+    private fun Observable<Message>.takeWhileMessagesInTimeout() : Observable<Message> =
+        takeWhile {
+            checkOnMessageTimeout(it.metadata.timestamp).also { continueObservation ->
+                hasMessagesInTimeoutInterval = hasMessagesInTimeoutInterval or continueObservation
+                if (!continueObservation) {
+                    streamCompletedState = State.MESSAGE_TIMEOUT
+                }
+            }
+        }
+
+    private fun calculateCheckpointTimeout(timestamp: Timestamp?, messageTimeout: Long?): Timestamp? =
+        if (timestamp == null || messageTimeout == null) {
```

Comment:
Shouldn't we alert if the `messageTimeout` is set but the checkpoint does not have `timestamp`?